rust/noosphere/src/ffi/petname.rs

```diff
@@ -1,47 +1,173 @@
 #![allow(unused_variables)]
 
-use safer_ffi::prelude::*;
+use std::str::FromStr;
 
-use crate::ffi::NsError;
+use anyhow::anyhow;
+use cid::Cid;
+use noosphere_sphere::{SpherePetnameRead, SpherePetnameWrite, SphereWalker};
+use safer_ffi::{char_p::InvalidNulTerminator, prelude::*};
 
-use super::NsNoosphereContext;
+use crate::ffi::{NsError, TryOrInitialize};
+
+use super::{NsNoosphere, NsSphere};
 
 #[ffi_export]
-/// Get the DID that is assigned to the provided petname; note that the DID is
-/// the ID of the sphere, but in order to read the sphere you must resolve the
-/// DID to a CID, which tells you the version of the sphere to read.
+/// Get the sphere identity - a DID - assigned to a given petname of a sphere.
+/// Note that this call will produce an error if no address has been assigned to
+/// the name.
 pub fn ns_sphere_petname_get(
-    noosphere: &NsNoosphereContext,
-    sphere_identity: char_p::Ref<'_>,
+    noosphere: &NsNoosphere,
+    sphere: &NsSphere,
     petname: char_p::Ref<'_>,
     error_out: Option<Out<'_, repr_c::Box<NsError>>>,
 ) -> Option<char_p::Box> {
-    todo!();
+    error_out.try_or_initialize(|| {
+        noosphere.async_runtime().block_on(async {
+            sphere
+                .inner()
+                .get_petname(petname.to_str())
+                .await?
+                .ok_or_else(|| anyhow!("No petname '{}' has been set", petname.to_str()))?
+                .to_string()
+                .try_into()
+                .map_err(|error: InvalidNulTerminator<String>| anyhow!(error).into())
+        })
+    })
 }
 
 #[ffi_export]
-/// Assign a DID to a petname. This will overwrite a petname entry if one already exists
-/// with the given name (and reset the resolved CID, if any).
+/// Assign a petname to a sphere identity (a DID). This will overwrite the
+/// petname with the new sphere identity if it has been set previously. When a
+/// new address is assigned to a petname, the petname will be reset to an
+/// unresolved state. Make sure to invoke sync after setting new petnames, so
+/// that new sphere identities can be resolved by the gateway (if one is
+/// configured).
```

Comment:
nit: In `ns_sphere_petname_get` and `ns_sphere_petname_set` (line 3), a sphere identity/DID is *assigned to* a petname. In `ns_sphere_petname_set` line 1, a petname is *assigned to* an identity/DID. While I do think of the structure as assigning a petname to identity, the inverse relationship is true and more common in comments (e.g. that the petname, not DID, is the key)

Reply:
I'm happy to standardize on whatever order makes the most sense.

Reply:
My intuition diverges here depending on whether you're thinking from the UI layer or the protocol layer. 

Within Noosphere it seems correct to say that you _assign_ a petname as a label for a DID, where-as in Subconscious it feels natural to say you create an address book entry by entering a petname and _assigning_ a DID to it (though we can flip that around through UI design if we want to).

Given that this is a Noosphere-level abstraction I'd lean towards "a petname is assigned to a DID", personally.